### PR TITLE
Udeng 2977 add progress bar to the docker icons

### PR DESCRIPTION
**Pull request by sergio-costas**
This PR shows a progress bar in the docker, over the icon of any snap being refreshed.

### CODE REVIEW

**Comment by sergio-costas** [comment]
@robert-ancell This is ready for review.

File: src/sdi-refresh-monitor.c [hunk 1]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
                                 g_object_ref(data->self));
 }
 
+static gboolean status_is_done(const gchar *status) {
+  gboolean done = g_str_equal(status, "Done") | g_str_equal(status, "Abort") |
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:256]: `|` -> `||`

File: src/sdi-refresh-monitor.c [hunk 2]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
...
+                  g_str_equal(status, "Error") | g_str_equal(status, "Hold") |
+                  g_str_equal(status, "Wait");
+  return done;
+}
+
+static void update_inhibited_snaps(SdiRefreshMonitor *self, SnapdChange *change,
+                                   gboolean done, gboolean hold) {
+  SnapdAutorefreshChangeData *changeData =
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:264]: changeData -> change_data

File: src/sdi-refresh-monitor.c [hunk 3]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
...
+      SNAPD_AUTOREFRESH_CHANGE_DATA(snapd_change_get_data(change));
+
+  if (changeData == NULL) {
+    return;
+  }
+
+  GStrv snapNames = snapd_autorefresh_change_data_get_snap_names(changeData);
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:271]: snapNames -> snap_names

File: src/sdi-refresh-monitor.c [hunk 4]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
...
+    // Only show progress bar if that snap shown an 'inhibited' notification
+    if (snap == NULL)
+      continue;
+
+    if (!sdi_snap_get_inhibited(snap))
+      continue;
+
+    if (done || hold) {
+      remove_snap(self, snap);
+      SnapRefreshData *data = g_malloc0(sizeof(SnapRefreshData));
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:283]: Would be a little cleaner with a `snap_refresh_data_new()`.

File: src/sdi-refresh-monitor.c [hunk 5]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
...
+    if (sdi_snap_get_hidden(snap) || sdi_snap_get_manually_hidden(snap)) {
+      continue;
+    }
+
+    g_autoptr(SdiRefreshDialog) dialog = sdi_snap_get_dialog(snap);
+    if (dialog == NULL) {
+      // if there's no dialog, get the data for this snap and create
+      // it avoid refresh notifications while the progress dialog is shown
+      sdi_snap_set_ignored(snap, TRUE);
+      SnapRefreshData *data = g_malloc0(sizeof(SnapRefreshData));
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:302]: As before, `snap_refresh_data_new()`

File: src/sdi-refresh-monitor.c [hunk 6]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
...
+      data->self = g_object_ref(self);
+      snapd_client_get_snap_async(self->client, *p, NULL,
+                                  begin_application_refresh, data);
+      continue;
+    }
+
+    // if there's already a dialog for this snap, just refresh the
+    // progress bar
+    GPtrArray *tasks = snapd_change_get_tasks(change);
+    gint done = 0;
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:313]: gint -> size_t / gsize.

File: src/sdi-refresh-monitor.c [hunk 7]
```diff
@@ -224,6 +252,161 @@ static void refresh_change(gpointer p) {
...
+  task_data->old_progress = progress;
+  g_autoptr(GVariantBuilder) builder =
+      g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+  g_variant_builder_add(builder, "{sv}", "progress",
+                        g_variant_new_double(progress));
+  g_variant_builder_add(builder, "{sv}", "progress-visible",
+                        g_variant_new_boolean(!task_data->done));
+  // values has a floating reference, so it's consumed when calling
+  // EMIT_UPDATE()
+  GVariant *values = g_variant_builder_end(builder);
```
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:351]: But what if `task_data->desktop_files->len == 0` won't it leak then? Better to take the reference and ensure it's always cleaned up.

**Review by robert-ancell — APPROVED** [review]
Some small stuff but otherwise looks good!

File: src/sdi-refresh-monitor.c (hunk 1 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:256]: Ops!!!! You are right.

Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:256]: Done

File: src/sdi-refresh-monitor.c (hunk 2 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:264]: Done.

File: src/sdi-refresh-monitor.c (hunk 3 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:271]: Done.

File: src/sdi-refresh-monitor.c (hunk 4 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:283]: You are right. Done.

File: src/sdi-refresh-monitor.c (hunk 5 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:302]: Also done.

File: src/sdi-refresh-monitor.c (hunk 6 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:313]: Done.

File: src/sdi-refresh-monitor.c (hunk 7 above)
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:351]: Good point. Done.

**Pull request merged**